### DISCUSSION TITLE
Is this a recommended pattern when building a Design System with StyleX.

### DISCUSSION THREAD

**Discussion by thisizkp** [facebook/stylex#329] [Q&A] [answered]
Hey Team. Hope you are doing well. Loving Stylex and thanks for all the amazing work you all have been doing.

We at work have started building a design system. We would like to use React-Aria & StyleX to achieve this. We've built our very first component so far (heading component). Would like to understand if we had missed anything or If this is the recommended pattern. Here's the code for the heading component (Intentionally restricting the variants).

```tsx
import { Heading as RACHeading } from "react-aria-components";
import * as stylex from "@stylexjs/stylex";
import { fontFamily, fontSize, fontWeight } from "../theme/tokens.stylex";

export type HeadingProps = {
  variant?: "h1" | "h2" | "h3" | "h4" | "h5";
  fontWeight?: "normal" | "medium" | "semibold" | "bold";
  children: React.ReactNode;
};

export function Heading({
  variant = "h3",
  fontWeight = "normal",
  children,
}: HeadingProps) {
  const level = parseInt(variant[1], 10);
  return (
    <RACHeading
      level={level}
      {...stylex.props(styles.base, styles[variant], styles[fontWeight])}
    >
      {children}
    </RACHeading>
  );
}

const styles = stylex.create({
  base: {
    fontFamily: fontFamily.base,
  },
  h1: {
    fontSize: fontSize["4xl"],
  },
  h2: {
    fontSize: fontSize["3xl"],
  },
  h3: {
    fontSize: fontSize["2xl"],
  },
  h4: {
    fontSize: fontSize.xl,
  },
  h5: {
    fontSize: fontSize.lg,
  },
  normal: {
    fontWeight: fontWeight.normal,
  },
  medium: {
    fontWeight: fontWeight.medium,
  },
  semibold: {
    fontWeight: fontWeight.semibold,
  },
  bold: {
    fontWeight: fontWeight.bold,
  },
})
```

I have created a separate tokens file which has tokens related to font family, size & weights. Of course it is readable but verbose too, so was thinking if on the right track or not. Thanks.

**Comment by nmn** [comment, marked as answer]
Everything looks good here!

If you're looking for any possible improvements, I would suggest splitting, the variants and fontWeights into separate `stylex.create` constants which will let you use something like this:

```tsx
stylex.props(styles.base, variants[variant], weights[fontWeight])
```

IMO, this improves the readability a bit furthers.

**Reply by thisizkp** [reply]
API surface layer is min (which is great), but then confusion around if I'm doing things right or wrong 😅.

Thanks. That was helpful.

**Reply by thisizkp** [reply]
@nmn Do I need to worry about reusing these things like `weights` for other components, or go ahead and just create one more where required (colocation) and let the compiler handle the deduping?

I'm guessing compiler does that for me, but want to make sure about it.

**Reply by nmn** [reply]
> or go ahead and just create one more where required (colocation) and let the compiler handle the deduping?

Go ahead and just create one more where required (colocation) and let the compiler handle the deduping!

----

Unless you *want* to share styles for some architectural reason such as:
* You have multiple components that should always use the same styles while having different implementations
* You just want to!
